lib/dynamoid/transaction/update.rb

```diff
@@ -0,0 +1,69 @@
+require_relative 'action'
+
+module Dynamoid
+  class Transaction
+    class Update < Action
+
+      def initialize(model_or_model_class, id, attributes, options = {})
+        model = model_or_model_class.is_a?(Dynamoid::Document) ?
+                       model_or_model_class : model_or_model_class.find(id, consistent_read: true)
+        super(model, attributes, options)
+      end
+
+      def run_callbacks
+        model.run_callbacks(:save) do
+          model.run_callbacks(:update) do
+            model.run_callbacks(:validate) do
+              yield if block_given?
+            end
+          end
+        end
+      end
+
+      def to_h
+        changes = model.changes.map{|k,v| [k.to_sym,v[1]]}.to_h # hash of dirty attributes
+        changes = changes.merge attributes if attributes.present?
+        changes = add_timestamps(changes, skip_created_at: true)
+        item = Dynamoid::Dumping.dump_attributes(changes, model_class.attributes)
+
+        # set 'key' that is used to look up record for updating
+        key = {model_class.hash_key => model.hash_key}
+        key[model_class.range_key] = model.range_key if model_class.range_key?
+
+        # e.g. "SET #updated_at = :updated_at ADD record_count :i"
+        item_keys = item.keys
+        update_expression = options[:update_expression] || ("SET " + item_keys.each_with_index.map{ |k, i| "##{k} = :_s#{i}" }.join(", "))
+
+        # e.g. {":updated_at" => 1645453.234, ":i" => 1}
+        expression_attribute_values = item_keys.each_with_index.map{|k,i| [":_s#{i}",item[k]]}.to_h
+
+        if options[:add]
+          add_keys = options[:add].keys
+          update_expression += " ADD " + add_keys.each_with_index.map{|k,i| "#{k} :_a#{i}"}.join(", ")
+          add_keys.each_with_index{|k,i| expression_attribute_values[":_a#{i}"] = options[:add][k] }
+        end
```

Comment:
I would postpone supporting `update_expression` and `add` options until they are asked for.